versioned/storage/cleanup/src/test/java/org/projectnessie/versioned/storage/cleanup/TestCutHistory.java

```diff
@@ -0,0 +1,310 @@
+
+    // Cut point: C1
+    //           /- G3 ---------\
+    //          /- E1 - E2 -\    \ /- F1
+    //  root - C0 --------- C1 - C2 - B1... Bnn
+    //                       \- D1 - D2
+    var main =
+        requireNonNull(repositoryLogic(persist).fetchRepositoryDescription()).defaultBranchName();
+    VersionStore store = new VersionStoreImpl(persist);
+    var root = commit(store, main, List.of("root"));
+
+    Set<String> c0Keys =
+        IntStream.rangeClosed(1, numExtraKeys).mapToObj(i -> "k" + i).collect(Collectors.toSet());
+    var c0 = commit(store, main, c0Keys); // force index stripes to still out
+
+    var branchE = store.create(BranchName.of("e"), Optional.of(c0)).getNamedRef().getName();
+    var e1 = commit(store, branchE, List.of("e1"));
+    var e2 = commit(store, branchE, List.of("e2"));
+
+    var branchG = store.create(BranchName.of("g"), Optional.of(c0)).getNamedRef().getName();
+    var g1 = commit(store, branchG, List.of("g1"));
+
+    var c1 =
+        store
+            .merge(
+                VersionStore.MergeOp.builder()
+                    .expectedHash(Optional.of(c0))
+                    .fromRef(BranchName.of(branchE))
+                    .fromHash(e2)
+                    .toBranch(BranchName.of(main))
+                    .build())
+            .getResultantTargetHash();
+
+    var branchD = store.create(BranchName.of("d"), Optional.ofNullable(c1)).getNamedRef().getName();
+    var d1 = commit(store, branchD, List.of("d1"));
+    var d2 = commit(store, branchD, List.of("d2"));
+
+    var c2 =
+        store
+            .merge(
+                VersionStore.MergeOp.builder()
+                    .expectedHash(Optional.of(c0))
+                    .fromRef(BranchName.of(branchG))
+                    .fromHash(g1)
+                    .toBranch(BranchName.of(main))
+                    .build())
+            .getResultantTargetHash();
+
+    var branchF = store.create(BranchName.of("f"), Optional.ofNullable(c2)).getNamedRef().getName();
+    var f1 = commit(store, branchF, List.of("f1"));
+
+    List<String> keysB = new ArrayList<>();
+    List<Hash> hashesB = new ArrayList<>();
+    for (int i = 0; i < numExtraKeys; i++) {
+      String k = "b" + i;
+      keysB.add(k);
+      hashesB.add(commit(store, main, List.of(k)));
+    }
+
+    var cleanup = createCleanup(CleanupParams.builder().build());
+    CutHistoryParams ctx = cleanup.buildCutHistoryParams(persist, hashToObjId(requireNonNull(c1)));
+    CutHistory cutHistory = cleanup.createCutHistory(ctx);
+    CutHistoryScanResult identifiedCommits = cutHistory.identifyAffectedCommits();
+
+    UpdateParentsResult rewriteResult = cutHistory.rewriteParents(identifiedCommits);
+    soft.assertThat(rewriteResult.failures()).isEmpty();
+
+    UpdateParentsResult cutResult = cutHistory.cutHistory();
+    soft.assertThat(cutResult.failures()).isEmpty();
+
+    PaginationIterator<Commit> it = store.getCommits(c1, true);
```

Comment:
Can you add a call to purge unreferecned objects? (Maybe using a parameterized flag whether purging happens)

Reply:
This depends on #10083... Putting this PR to "draft" state for now.

Reply:
added tests with purging